tools/reduce.py

```diff
@@ -0,0 +1,303 @@
+            tempdir = tempfile.mkdtemp(prefix="scip-clang-reduce-")
+            exitstack.callback(
+                lambda: logging.info(
+                    f"scip-clang temporary directory persisted at {tempdir}"
+                )
+            )
+            tempdir = pathlib.Path(tempdir)
+            self.entry.directory = tempdir
+
+            shutil.copyfile(absolute_start_tu_path, tempdir.joinpath(start_tu_path))
+            compdb = tempdir.joinpath("compile_commands.json")
+            with open(compdb, "w") as compdb_file:
+                json.dump([self.entry.to_dict()], compdb_file)
+
+            run_sh = tempdir.joinpath("run.sh")
+            with open(run_sh, "w") as run_sh_file:
+                assert self.scip_clang.is_absolute()
+                run_sh_file.write(
+                    textwrap.dedent(
+                        f"""\
+                    #!/usr/bin/env bash
+                    {self.scip_clang} --worker-mode=compdb --compdb-path={compdb} > out.log 2>&1
+                    set -e
+                    grep -E '{self.scip_clang_output_pattern.pattern}' out.log
+                    """
+                    )
+                )
+            run_sh.chmod(run_sh.stat().st_mode | stat.S_IEXEC)
+
+            result = subprocess.run(
+                ["creduce", str(run_sh), str(start_tu_path)] + self.extra_args,
+                cwd=tempdir,
+            )
+            if result.returncode != 0:
+                sys.exit(result.returncode)
+
+    def fix_paths(self):
+        """
+        Adjusts the paths in the argument list so that the compilation database
+        entry can be relocated correctly to a temporary directory.
+
+        Post-condition:
+        - Include paths, such as using -I and -isysroot use absolute paths
+        - The path for the main TU has a relative path
+
+        Creduce creates temporary directories, so we want all the paths to
+        work properly from those directories too.
+        """
+        original_path = self.entry.filepath
+        if self.entry.filepath.is_absolute():
+            self.entry.filepath = self.entry.filepath.relative_to(self.project_root)
+        tu_path_str = str(self.entry.filepath)
+
+        def fix_path(maybe_path: str) -> str:
+            if os.path.isabs(maybe_path):
+                return maybe_path
+            abs_path = self.entry.directory.joinpath(maybe_path)
+            if abs_path.exists():
+                return str(abs_path)
+            # Maybe it wasn't a path argument after all...
+            return maybe_path
+
+        for i, arg in enumerate(self.entry.arguments):
+            if tu_path_str in arg:
+                if os.path.isabs(arg):
+                    self.entry.arguments[i] = tu_path_str
+            else:
+                match = re.match(self.path_flag_pattern, arg)
+                if match is None:
+                    continue
+                if match.group("path") != "":
```

Comment:
Empty string is false, so be brief and omit != "" (unless there's some type checker we are using that hates brevity...)